File: frontend/src/api/ledgerImport.ts
import { api } from './index';

// ---------- 导入配置 ----------

export interface ImportConfig {
  template_name: string;
  max_rows: number;
  warn_rows: number;
  max_rows_hard_limit: number;
  name_field?: string | null;
  unique_keys: string[];
  identity_mapping: Record<string, string>;
  field_order: string[];
}

export async function listImportConfigs() {
  const response = await api.get('/ledgers/import-configs');
  return response.data as { configs: ImportConfig[] };
}

export async function reloadImportConfigs() {
  const response = await api.post('/ledgers/import-configs/reload');
  return response.data as { message: string; count: number };
}

// ---------- 导入模板下载 ----------

export async function downloadImportTemplate(templateId: number) {
  const response = await api.get(`/ledgers/import-template/${templateId}`, {
    responseType: 'blob',
    headers: {
      Accept: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    },
  });
  return response.data as Blob;
}

// ---------- 预校验 / 提交 ----------

export interface ImportIssue {
  row: number;
  type: string;
  field?: string | null;
  raw?: unknown;
  reason: string;
}

export interface CleanablePreview {
  row: number;
  field: string;
  raw: unknown;
  clean: unknown;
}

export interface ImportValidationReport {
  template_id: number;
  template_name: string;
  preview_batch_id: string;
  total_rows: number;
  importable_count: number;
  cleanable_count: number;
  issues: ImportIssue[];
  cleanable_previews: CleanablePreview[];
  row_limit_warning?: string | null;
}

export async function validateImport(
  templateId: number,
  file: File,
  teamId?: number,
) {
  const formData = new FormData();
  formData.append('template_id', String(templateId));
  formData.append('file', file);
  if (teamId !== undefined && teamId !== null) {
    formData.append('team_id', String(teamId));
  }
  const response = await api.post('/ledgers/import/validate', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return response.data as ImportValidationReport;
}

export interface ImportFailedItem {
  row: number;
  field?: string | null;
  raw?: unknown;
  reason: string;
}

export interface ImportCommitResult {
  batch_id: string;
  template_id: number;
  template_name: string;
  success_count: number;
  failed_count: number;
  elapsed_seconds: number;
  created_ledger_ids: number[];
  failed_items: ImportFailedItem[];
  failures_download_url?: string | null;
}

export async function commitImport(
  templateId: number,
  file: File,
  teamId?: number,
) {
  const formData = new FormData();
  formData.append('template_id', String(templateId));
  formData.append('file', file);
  if (teamId !== undefined && teamId !== null) {
    formData.append('team_id', String(teamId));
  }
  const response = await api.post('/ledgers/import/commit', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return response.data as ImportCommitResult;
}

// ---------- 导入历史 ----------

export interface ImportHistoryItem {
  batch_id: string;
  template_id: number;
  template_name: string;
  imported_by_id: number;
  imported_by_name?: string | null;
  imported_at: string;
  total_rows: number;
  success_count: number;
  failed_count: number;
  filename?: string | null;
}

export interface ImportHistoryList {
  total: number;
  items: ImportHistoryItem[];
}

export async function listImportHistory(skip = 0, limit = 20) {
  const response = await api.get(`/ledgers/import/history?skip=${skip}&limit=${limit}`);
  return response.data as ImportHistoryList;
}

export interface ImportHistoryDetail extends ImportHistoryItem {
  created_ledger_ids: number[];
  failed_items?: ImportFailedItem[] | null;
}

export async function getImportHistoryDetail(batchId: string) {
  const response = await api.get(`/ledgers/import/history/${batchId}`);
  return response.data as ImportHistoryDetail;
}

export async function downloadImportFailures(batchId: string) {
  const response = await api.get(`/ledgers/import/history/${batchId}/failures`, {
    responseType: 'blob',
    headers: {
      Accept: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    },
  });
  return response.data as Blob;
}

File: backend/app/schemas/ledger_import.py
"""台账导入相关 Pydantic Schemas"""
from typing import Any, Dict, List, Optional

from pydantic import BaseModel


# ---------- 导入配置 ----------

class ImportConfig(BaseModel):
    """单条模板导入配置(来自 ledger_import_config.json)"""
    template_name: str
    max_rows: int = 2000
    warn_rows: int = 5000
    max_rows_hard_limit: int = 5000
    name_field: Optional[str] = None
    unique_keys: List[str] = ["name"]
    identity_mapping: Dict[str, str] = {}
    field_order: List[str] = []


class ImportConfigList(BaseModel):
    configs: List[ImportConfig]


# ---------- 校验报告 ----------

class ImportIssue(BaseModel):
    row: int                     # Excel 行号(从 1 计,含表头偏移)
    type: str                    # missing_required / duplicate / team_not_found /
                                 # user_not_found / uncleanable_value / type_mismatch /
                                 # row_limit_exceeded
    field: Optional[str] = None  # 出问题的字段 label
    raw: Optional[Any] = None    # 原始值
    reason: str


class CleanablePreview(BaseModel):
    row: int
    field: str
    raw: Any
    clean: Any


class ImportValidationReport(BaseModel):
    template_id: int
    template_name: str
    preview_batch_id: str
    total_rows: int
    importable_count: int
    cleanable_count: int
    issues: List[ImportIssue] = []
    cleanable_previews: List[CleanablePreview] = []
    row_limit_warning: Optional[str] = None


# ---------- 提交结果 ----------

class ImportFailedItem(BaseModel):
    row: int
    field: Optional[str] = None
    raw: Optional[Any] = None
    reason: str


class ImportCommitResult(BaseModel):
    batch_id: str
    template_id: int
    template_name: str
    success_count: int
    failed_count: int
    elapsed_seconds: float
    created_ledger_ids: List[int] = []
    failed_items: List[ImportFailedItem] = []
    failures_download_url: Optional[str] = None
